services/s3/itest/s3_test.go

```diff
@@ -793,3 +793,175 @@ func TestHead(t *testing.T) {
 		t.Fatal("unexpected content length")
 	}
 }
+
+func TestPutObjectIfNoneMatch(t *testing.T) {
+	ctx := context.Background()
+	client, srv := makeClientServerPair()
+	defer srv.Shutdown(ctx)
+
+	ifNoneMatchWildcard := "*"
+	key := "test-key"
+	_, err := client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket:      &bucket,
+		Key:         &key,
+		Body:        strings.NewReader("hello"),
+		IfNoneMatch: &ifNoneMatchWildcard,
+	})
+	// First put should succeed
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Second put must fail since an object already exists with that key
+	_, expectedErr := client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket:      &bucket,
+		Key:         &key,
+		Body:        strings.NewReader("world"),
+		IfNoneMatch: &ifNoneMatchWildcard,
+	})
+
+	if expectedErr == nil {
+		t.Fatal(expectedErr)
```

Comment:
replacing this with `t.Fatal("expected error")` since `t.Fatal(nil)` isn't super helpful :)